drivers: adc: add dt-spec helper functions

gmarull opened the pull request:
This patch adds the following dt-spec helpers:

- `adc_channel_setup_dt()`
- `adc_raw_to_millivolts_dt()`
- `adc_sequence_init_dt()`

The objective of these functions is to reduce application boilerplate
when Devicetree is used to configure the ADC.

`sampled/drivers/adc` has been refactored to make use of these helpers. It now requires usage of Devicetree to configure channels, making it simpler and more portable.

gmarull committed:
drivers: adc: add dt-spec helper functions

This patch adds the following dt-spec helpers:

- adc_channel_setup_dt()
- adc_raw_to_millivolts_dt()
- adc_sequence_init_dt()

The objective of these functions is to reduce application boilerplate
when Devicetree is used to configure the ADC.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: adc: remove redundant controller labels

The controller name can be obtained from its device name field instead.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: adc: configure ADC channel/s in Devicetree

It is now possible to configure ADC channels in Devicetree. This makes
code more portable, since there is no need to hardcode hardware specific
bits in the C sources.

This patch adds channel/s configuration for all support sample overlays.
Hardcoded values have been used:

- Resolution: 12-bit
- Gain: 1
- Reference: internal
- Acquisition time: default

Some nRF channels have been adjusted to include resolution as well.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: adc: rely on Devicetree channel configuration

Simplify sample by relying exclusively on channel data from Devicetree.
This change makes sample simpler and more portable at the same time.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: adc: minor fixes and cleanups

- Improve include list, so that file is self-contained
- Don't use an array type for the samples buffer, as its length is 1
- Improve some error messages
- Use PRId16/32 for printing fixed-width signed integers

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: adc: adjust README

- Inform about the requirement to configure ADC channels in Devicetree
- Fix sample output example

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on include/zephyr/drivers/adc.h:
```diff
@@ -402,44 +402,6 @@ struct adc_dt_spec {
  */
 #define ADC_DT_SPEC_INST_GET(inst) ADC_DT_SPEC_GET(DT_DRV_INST(inst))
 
-/**
```
note: has been moved because the dt-spec version requires calling a function defined below.

nandojve commented on include/zephyr/drivers/adc.h:
```diff
@@ -731,6 +706,118 @@ static inline uint16_t adc_ref_internal(const struct device *dev)
+ * in a struct adc_dt_spec.
+ *
+ * @param[in] spec ADC specification from Devicetree.
+ * @param[in,out] valp Pointer to the raw measurement value on input, and the
+ *                     corresponding millivolt value on successful conversion.
+ *                     If conversion fails the stored value is left unchanged.
+ *
```
I noted this is vertically aligned which is different from remaining documentation.

gmarull replied:
made consistent

nandojve commented on include/zephyr/drivers/adc.h:
```diff
@@ -731,6 +706,118 @@ static inline uint16_t adc_ref_internal(const struct device *dev)
+
+	resolution = spec->resolution;
+
+	/*
+	 * For differential channels, one bit less needs to be specified
+	 * for resolution to achieve correct conversion.
+	 */
+	if (spec->channel_cfg.differential) {
+		resolution -= 1U;
+	}
```
This not match with the documentation at lines 722~723. I mean, it "`needs to be`" or it "`may be`". Looking the code it seems that `needs to be` is the correct one.

gmarull replied:
@anangl could you comment on that? From an API perspective it's marked as "optional", but sample did this operation unconditionally.

anangl replied:
In my opinion this "may be" in the description of `adc_raw_to_millivolts` should be corrected, as it is not consistent with the description of `resolution`: https://github.com/zephyrproject-rtos/zephyr/blob/2c1ef5598cb809d665fbe4bf496725a33a0928ea/include/zephyr/drivers/adc.h#L519-L526 This -1 cannot be "optional".

nandojve commented on samples/drivers/adc/src/main.c:
```diff
@@ -146,16 +65,14 @@ void main(void)
+			err = adc_raw_to_millivolts_dt(&adc_channels[i],
+						       &val_mv);
+			if (err == 0) {
+				printk(" = %d mV", val_mv);
```
This is a sample, should be interesting to tell users that mV conversion was skip or it is not available for driver foo?

gmarull replied:
good point, improved

nandojve reviewed (changes requested):
Hi @gmarull ,

Nice change! Just few minors to double check.

henrikbrixandersen commented on samples/drivers/adc/src/main.c:
```diff
@@ -22,149 +29,58 @@ static const struct adc_dt_spec adc_channels[] = {
 {
 	int err;
-	int16_t sample_buffer[1];
+	int16_t buf;
```
Note: this code still makes an assumption about the data type used by the ADC (16-bit signed, but some ADCs use 32-bit data types). I have yet to come up with a good suggestion for how to deal with this. Ideas are most welcome.

gmarull replied:
I guess it could be deduced from resolution, right?

anangl commented on include/zephyr/drivers/adc.h:
```diff
@@ -644,6 +606,19 @@ static inline int z_impl_adc_channel_setup(const struct device *dev,
+ */
+static inline int adc_channel_setup_dt(const struct adc_dt_spec *spec)
+{
+	return adc_channel_setup(spec->dev, &spec->channel_cfg);
```
`spec->channel_cfg_dt_node_exists` should be checked before this call is done.

gmarull replied:
done, and documented return code

anangl commented on include/zephyr/drivers/adc.h:
```diff
@@ -731,6 +706,118 @@ static inline uint16_t adc_ref_internal(const struct device *dev)
+ * corresponding millivolt value on successful conversion. If conversion fails
+ * the stored value is left unchanged.
+ *
+ * @return A value from adc_raw_to_millivolts().
```
Return value `-ENOTSUP` should be also documented.

gmarull replied:
done

anangl commented on samples/drivers/adc/src/main.c:
```diff
@@ -22,149 +29,58 @@ static const struct adc_dt_spec adc_channels[] = {
-		configure_channel(&adc_channels[i], &vref_mv[i]);
+		err = adc_channel_setup_dt(&adc_channels[i]);
+		if (err < 0) {
+			printk("Could not setup channel (%d)\n", err);
```
Perhaps it would be nice to point out which channel could not be set up. Something like:
```suggestion
			printk("Could not setup channel #%d (%d)\n", i, err);
```

anangl commented on samples/drivers/adc/src/main.c:
```diff
@@ -22,149 +29,58 @@ static const struct adc_dt_spec adc_channels[] = {
 			     DT_SPEC_AND_COMMA)
 };
 
-#define LABEL_AND_COMMA(node_id, prop, idx) \
-	DT_LABEL(DT_IO_CHANNELS_CTLR_BY_IDX(node_id, idx)),
-
-/* Labels of ADC controllers referenced by the above io-channels. */
-static const char *const adc_labels[] = {
-	DT_FOREACH_PROP_ELEM(DT_PATH(zephyr_user), io_channels,
-			     LABEL_AND_COMMA)
-};
```
👍 Thanks a lot for simplifying this.

I don't know how I could miss this that I could just use `dev->name` instead of complicating things this way... 🤦

gmarull committed:
drivers: adc: clarify differential sampling resolution

When using differential sampling, resolution needs to be one less, not
"may be".

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

anangl commented on samples/drivers/adc/src/main.c:
```diff
@@ -22,149 +29,58 @@ static const struct adc_dt_spec adc_channels[] = {
+		for (size_t i = 0U; i < ARRAY_SIZE(adc_channels); i++) {
+			int32_t val_mv;
+
+			printk(" - %s, channel %d: ",
```
```suggestion
			printk("- %s, channel %d: ",
```
See a78e7c5c90d8cb31ec550ce289503bd698675585.
